File: packages/sdk/python/opentool/resources/tools.py
from __future__ import annotations

from typing import Any, Dict, List, Optional, Union

from opentool.http import HttpClient, AsyncHttpClient
from opentool.types import (
    Tool,
    ToolExecutionResult,
    ToolSearchResult,
    ToolSearchSummary,
    ToolDetails,
)


class ToolsResource:
    """Sync tool listing, provider browsing, search, and execution."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(self) -> List[Tool]:
        """List all tools in the registry (no auth required)."""
        data = self._http.get("/api/tools/")
        return [Tool.model_validate(t) for t in data["tools"]]

    def connected(self) -> List[Tool]:
        """List tools the authenticated user has connected."""
        data = self._http.get("/api/tools/connected")
        return [Tool.model_validate(t) for t in data["tools"]]

    def by_provider(self, provider: str) -> List[Tool]:
        """List tools for a specific provider."""
        data = self._http.get(f"/api/tools/{provider}")
        return [Tool.model_validate(t) for t in data["tools"]]

    def search(
        self,
        *,
        query: Optional[str] = None,
        provider: Optional[str] = None,
        category: Optional[str] = None,
        auth_type: Optional[str] = None,
        read_only: Optional[bool] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> Union[ToolSearchResult, ToolSearchSummary]:
        """
        Search and filter tools by keyword, provider, category, or capability.
        Returns a paginated list of matching tools.
        If no options are provided, returns a provider summary.
        """
        params: Dict[str, str] = {}
        if query:
            params["q"] = query
        if provider:
            params["provider"] = provider
        if category:
            params["category"] = category
        if auth_type:
            params["auth_type"] = auth_type
        if read_only is not None:
            params["read_only"] = str(read_only).lower()
        if limit is not None:
            params["limit"] = str(limit)
        if offset is not None:
            params["offset"] = str(offset)

        qs = "&".join(f"{k}={v}" for k, v in params.items())
        path = f"/api/tools/search?{qs}" if qs else "/api/tools/search"
        data = self._http.get(path)

        if "providers" in data:
            return ToolSearchSummary.model_validate(data)
        return ToolSearchResult.model_validate(data)

    def details(self, tool_id: str) -> ToolDetails:
        """
        Get full schema and metadata for a specific tool.
        Fetches the tool list and finds the matching tool.
        """
        data = self._http.get("/api/tools/")
        for t in data["tools"]:
            if t["id"] == tool_id:
                return ToolDetails.model_validate(t)
        raise ValueError(f'Tool "{tool_id}" not found')

    def execute(
        self, tool_id: str, args: Dict[str, Any] | None = None
    ) -> ToolExecutionResult:
        """
        Execute a tool via the MCP JSON-RPC endpoint.

        Example::

            result = client.tools.execute("github_create_issue", {
                "owner": "user",
                "repo": "my-repo",
                "title": "Bug report",
            })
        """
        rpc_response = self._http.post(
            "/mcp",
            {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "tools/call",
                "params": {"name": tool_id, "arguments": args or {}},
            },
        )
        if "error" in rpc_response:
            raise RuntimeError(
                f"Tool execution failed: {rpc_response['error']['message']}"
            )
        return ToolExecutionResult.model_validate(rpc_response["result"])


class AsyncToolsResource:
    """Async tool listing, provider browsing, search, and execution."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def list(self) -> List[Tool]:
        """List all tools in the registry (no auth required)."""
        data = await self._http.get("/api/tools/")
        return [Tool.model_validate(t) for t in data["tools"]]

    async def connected(self) -> List[Tool]:
        """List tools the authenticated user has connected."""
        data = await self._http.get("/api/tools/connected")
        return [Tool.model_validate(t) for t in data["tools"]]

    async def by_provider(self, provider: str) -> List[Tool]:
        """List tools for a specific provider."""
        data = await self._http.get(f"/api/tools/{provider}")
        return [Tool.model_validate(t) for t in data["tools"]]

    async def search(
        self,
        *,
        query: Optional[str] = None,
        provider: Optional[str] = None,
        category: Optional[str] = None,
        auth_type: Optional[str] = None,
        read_only: Optional[bool] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> Union[ToolSearchResult, ToolSearchSummary]:
        """Search and filter tools."""
        params: Dict[str, str] = {}
        if query:
            params["q"] = query
        if provider:
            params["provider"] = provider
        if category:
            params["category"] = category
        if auth_type:
            params["auth_type"] = auth_type
        if read_only is not None:
            params["read_only"] = str(read_only).lower()
        if limit is not None:
            params["limit"] = str(limit)
        if offset is not None:
            params["offset"] = str(offset)

        qs = "&".join(f"{k}={v}" for k, v in params.items())
        path = f"/api/tools/search?{qs}" if qs else "/api/tools/search"
        data = await self._http.get(path)

        if "providers" in data:
            return ToolSearchSummary.model_validate(data)
        return ToolSearchResult.model_validate(data)

    async def details(self, tool_id: str) -> ToolDetails:
        """Get full schema and metadata for a specific tool."""
        data = await self._http.get("/api/tools/")
        for t in data["tools"]:
            if t["id"] == tool_id:
                return ToolDetails.model_validate(t)
        raise ValueError(f'Tool "{tool_id}" not found')

    async def execute(
        self, tool_id: str, args: Dict[str, Any] | None = None
    ) -> ToolExecutionResult:
        """Execute a tool via the MCP JSON-RPC endpoint."""
        rpc_response = await self._http.post(
            "/mcp",
            {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "tools/call",
                "params": {"name": tool_id, "arguments": args or {}},
            },
        )
        if "error" in rpc_response:
            raise RuntimeError(
                f"Tool execution failed: {rpc_response['error']['message']}"
            )
        return ToolExecutionResult.model_validate(rpc_response["result"])

File: packages/sdk/ts/src/resources/tools.ts
import { HttpClient } from '../http'
import type {
  Tool,
  ToolList,
  ToolExecutionResult,
  ToolSearchOptions,
  ToolSearchResult,
  ToolSearchSummary,
  ToolDetails,
} from '../types'

export class ToolsResource {
  constructor(private http: HttpClient) {}

  /** List all tools in the registry (no auth required). */
  async list(): Promise<Tool[]> {
    const res = await this.http.get<ToolList>('/api/tools/')
    return res.tools
  }

  /** List tools the authenticated user has connected. */
  async connected(): Promise<Tool[]> {
    const res = await this.http.get<ToolList>('/api/tools/connected')
    return res.tools
  }

  /** List tools for a specific provider. */
  async byProvider(provider: string): Promise<Tool[]> {
    const res = await this.http.get<ToolList>(`/api/tools/${provider}`)
    return res.tools
  }

  /**
   * Search and filter tools by keyword, provider, category, or capability.
   * Returns a paginated list of matching tools.
   * If no options are provided, returns a provider summary.
   */
  async search(options: ToolSearchOptions = {}): Promise<ToolSearchResult | ToolSearchSummary> {
    const params = new URLSearchParams()
    if (options.query) params.set('q', options.query)
    if (options.provider) params.set('provider', options.provider)
    if (options.category) params.set('category', options.category)
    if (options.authType) params.set('auth_type', options.authType)
    if (options.readOnly !== undefined) params.set('read_only', String(options.readOnly))
    if (options.limit !== undefined) params.set('limit', String(options.limit))
    if (options.offset !== undefined) params.set('offset', String(options.offset))

    const qs = params.toString()
    return this.http.get<ToolSearchResult | ToolSearchSummary>(
      `/api/tools/search${qs ? `?${qs}` : ''}`,
    )
  }

  /**
   * Get full schema and metadata for a specific tool by its ID.
   * Calls the meta_get_tool_details MCP tool under the hood.
   */
  async details(toolId: string): Promise<ToolDetails> {
    // Use the REST tool search to get tool details by exact ID
    const list = await this.http.get<ToolList>(`/api/tools/`)
    const tool = list.tools.find((t) => t.id === toolId)
    if (!tool) {
      throw new Error(`Tool "${toolId}" not found`)
    }
    return tool as unknown as ToolDetails
  }

  /**
   * Execute a tool via the MCP JSON-RPC endpoint.
   */
  async execute(toolId: string, args: Record<string, unknown> = {}): Promise<ToolExecutionResult> {
    const rpcResponse = await this.http.post<{
      jsonrpc: string
      id: number
      result?: ToolExecutionResult
      error?: { code: number; message: string }
    }>('/mcp', {
      jsonrpc: '2.0',
      id: 1,
      method: 'tools/call',
      params: { name: toolId, arguments: args },
    })

    if (rpcResponse.error) {
      throw new Error(`Tool execution failed: ${rpcResponse.error.message}`)
    }

    return rpcResponse.result!
  }
}
